### DISCUSSION TITLE
Kit Script Layouts

### DISCUSSION THREAD

**Discussion by zzeleznick** [johnlindquist/kit#231] [Ideas]
### Background

One of my favorite parts about Script Kit is that it eliminates > 80% of the boilerplate required to spin up a simple app. The aha moment for me was being able to build a pokedex app in strictly < 40 lines of code (see #79).

I love that apps are essentially [userScripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/userScripts/Working_with_userScripts) bundled within in a native app via Electron. The developer productivity feels unreal, and creating the N+1 'th project is just opening up a new file vs. setting up a new Xcode or Electron project.

### Challenge

With the change of `html` behavior to span the full choice element, the layout of the Github discussions example (see #103) broke and I was thinking that it would be nice for developer ergonomics to enable passing in layout options to the `arg` function.

Thinking out loud, the majority of apps contain a custom list view, and whether building for iOS or Android, there are a set of common design patterns with layout styles.

### Common List View Layouts
<table>
<tr>
<td width="500" align="center">
<img width="250" alt="android-list-view" src="https://user-images.githubusercontent.com/5779832/116826342-637c0800-ab48-11eb-9e74-5eae423fa93a.png">
</td>
<td width="500">
<img width="500" alt="iOS Table cell view" src="https://user-images.githubusercontent.com/5779832/116825529-96240180-ab44-11eb-9975-cf495db0f1cd.png">
</td>
</tr>
<tr>
<td>
Android example with cell image on left
</td>
<td>
iOS example with cell image on right
</td>
</tr>
</table>

<hr>

### Proposal

Support conditional rendering of the choice elements by supplying a layout key from a preset list of options.

#### Example Layouts

<table>
<tr>
<td> <img width="500" alt="Script Kit Layouts" src="https://user-images.githubusercontent.com/5779832/116798809-468ff800-aaa8-11eb-91bd-ce977b6efbab.png">
</td>
<td>  <img width="500" alt="Script Kit Layouts 2" src="https://user-images.githubusercontent.com/5779832/116825454-4b09ee80-ab44-11eb-80b7-3bb9c7009510.png">
</td>
</tr>
<tr>
<td>
With `img` property set and `html` unset
</td>
<td>
With `html` property set and `img` unset
</td>
</tr>
</table>

#### Example

Display images on the left hand side of the choice cell element
```js
await arg("View List:", // prompt message
    [{name: "Title", description: "Subtitle", img: "https://place-hold.it/300.png"}], // choices
    {layout: "left"} // options
);
``` 

<hr>

### Notes

- This relates to theming (#84), but this emphasis here is on script-specific layout styles as opposed to global styles
- The keys (e.g. `left` / `right`) for the layout are for example purposes aren't intended to be final
- For other layout options, a grid view comes to mind which corresponds to [UICollectionViewFlowLayout ](https://developer.apple.com/documentation/uikit/uicollectionviewflowlayout) for iOS

**Comment by johnlindquist** [comment]
I think this is a wonderful idea!

For some context, [here's the css](https://github.com/johnlindquist/kitapp/blob/beta/src/components/button.tsx#L177) we would be changing, so it's a fairly straightforward addition using flexbox. I would have to move the `html` div back next to `img`, but that's totally fine. (Don't look at the code too long, I know I have _a lot_ of optimizing to do 😅 )

Two minor changes:

1. call it `className`
2. make it a property of choice array

```js
await arg("", [{name: "one", className:"right"}])
```

We'll have the default `left` and `right` classes you suggest, but we could also allow people to pass in a string to customize fonts, borders, background, etc if they want to emphasize a specific choice.

**Reply by zzeleznick** [reply]
Glad to hear it 🤩 !

Haha, the code looks pretty reasonable to me 😅 . I'm onboard for your recommend changes, too!

For clarity on the `html` behavior, I was thinking that the render logic would look like `html && ( name || description)` for the full-width case (to render only the `html` element where both `name` and `description` are unset) and re-adding the `html` div next to `img` as you mentioned.

In my head, I was thinking that `layout` would be one level of abstraction above `className`. 

The working (but less extensible) solution I thought of was creating a mapping from `{layout: "right"}` to Tailwind CSS `className = "flex-row-reverse"` (e.g. `flex-direction: row-reverse`) to swap the order of `choice.img` and the parent element of `choice.name` and `choice.description`. 

I'm onboard for (1) using `className` here and (2) making it a property of choice array to better support additional custom styling down the road.

Just to show some fleshed out examples of common layouts I was picturing

<table>
  <tr>
    <td rowspan="3"><img alt="Full Screen List Example" src="https://user-images.githubusercontent.com/5779832/116929067-bcb06e00-ac12-11eb-87de-83367f3358ae.png"></td>
    <td><img alt="Cats vs Dogs" src="https://user-images.githubusercontent.com/5779832/116929071-bf12c800-ac12-11eb-9761-08b73ebe38da.png"></td>
  </tr>
  <tr>
  <td>Example with <code>html</code> and <code>name</code> with <code>left</code> style </td>
</tr>
  <tr>
   <td><img alt="Meta Votes" src="https://user-images.githubusercontent.com/5779832/116929066-bc17d780-ac12-11eb-836a-dda803470a5b.png"></td>
  </tr>
<tr>
<td>Full-width <code>html</code> and <code>default</code> style. </td>  
<td>Example with <code>html</code>, <code>name</code>, and <code>description</code> with <code>left</code> style </td>
</tr>
</table>

For the full-width <code>html</code> case, you could say that there is in fact a <code>name</code>, and <code>description</code> implicitly encoded here, but they would likely need to be children rather than siblings of <code>html</code> in flexbox land.

**Reply by johnlindquist** [reply]
Sweet, thanks for the visuals.

I'll try some variations and see what works. We don't always have to conditionally render, since we could "display:none" with the classes.
